modules/dbfs/mysql.go

```diff
@@ -0,0 +1,452 @@
+)
+
+var mysqldb *mysqlConn
+
+var mysqldbName = "cc"
+
+type mysqlConn struct {
+	config config.ConnCfg
+	db     *sql.DB
+	dbname string
+}
+
+func openMySQLConn(dbName string) (*mysqlConn, error) {
+	if mysqldb != nil && mysqldb.db != nil {
+		err := mysqldb.db.Ping()
+		if err == nil && mysqldb.dbname == dbName {
+			return mysqldb, nil
+		}
+	}
+
+	if mysqldb == nil || mysqldb.config == (config.ConnCfg{}) {
+		mysqldb = new(mysqlConn)
+		configMap := config.GetConfig()
+		mysqldb.config = configMap.ConnectionConfig["MySQL"]
+	}
+
+	db, err := sql.Open("mysql", mysqldb.config.Username+":"+mysqldb.config.Password+"@tcp("+mysqldb.config.Host+":"+strconv.Itoa(int(mysqldb.config.Port))+")/"+dbName+"?timeout="+strconv.Itoa(int(mysqldb.config.Timeout))+"s"+"&parseTime=true")
+	if err != nil {
+		utils.LogOnError(err, "Unable to connect to MySQL")
+		return mysqldb, err
+	}
+
+	mysqldb.db = db
+	mysqldb.dbname = dbName
+	return mysqldb, nil
+}
+
+// CloseMySQL closes the MySQL db connection
+// YOU PROBABLY DON'T NEED TO RUN THIS EVER
+func CloseMySQL() error {
+	if mysqldb != nil && mysqldb.db != nil {
+		err := mysqldb.db.Close()
+		mysqldb = nil
+		return err
+	}
+	return ErrDbNotInitialized
+}
+
+/**
+STORED PROCEDURES
+*/
+
+// MySQLUserRegister registers a new user in MySQL
+func MySQLUserRegister(username string, pass string, email string, firstName string, lastName string) error {
+	mysql, err := openMySQLConn(mysqldbName)
+	if err != nil {
+		return err
+	}
+
+	result, err := mysql.db.Exec("CALL user_register(?,?,?,?,?)", username, pass, email, firstName, lastName)
+	if err != nil {
+		return err
+	}
+	numrows, err := result.RowsAffected()
+
+	if err != nil || numrows == 0 {
+		return ErrNoDbChange
+	}
+
+	return nil
+}
+
+// MySQLUserGetPass is used to get the key and hash of a stored password to verify that a value is correct
+func MySQLUserGetPass(username string) (password string, err error) {
+	mysql, err := openMySQLConn(mysqldbName)
+	if err != nil {
+		return "", err
+	}
+
+	rows, err := mysql.db.Query("CALL user_get_password(?)", username)
+	if err != nil {
+		return "", err
+	}
+
+	for rows.Next() {
+		err = rows.Scan(&password)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return password, nil
+}
+
+// MySQLUserDelete deletes a user from MySQL
+// unexported b/c not part of the official API
+func mySQLUserDelete(username string, pass string) error {
+	mysql, err := openMySQLConn(mysqldbName)
+	if err != nil {
+		return err
+	}
+
+	// FIXME: use MySQLUserGetPass to verify the password is correct before deleting
+
+	result, err := mysql.db.Exec("CALL user_delete(?)", username)
+	if err != nil {
+		return err
+	}
+	numrows, err := result.RowsAffected()
+
+	if err != nil || numrows == 0 {
+		return ErrNoDbChange
+	}
+
+	return nil
+}
+
+// MySQLUserLookup returns user information about a user with the username 'username'
+func MySQLUserLookup(username string) (firstname string, lastname string, email string, err error) {
```

Comment:
Should we make the return a model? it seems like we'd use this type of model quite often.